tests/Fixtures/TestBundle/Entity/HiddenIdentifierDummy.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Tests\Fixtures\TestBundle\Entity;
+
+use ApiPlatform\Metadata\ApiProperty;
+use ApiPlatform\Metadata\ApiResource;
+use ApiPlatform\Metadata\Get;
+use ApiPlatform\Metadata\Put;
+use Doctrine\ORM\Mapping\Column;
+use Doctrine\ORM\Mapping\Entity;
+use Doctrine\ORM\Mapping\Id;
+use Symfony\Component\Serializer\Annotation\Groups;
+
+#[ApiResource(
+    operations            : [new Get(), new Put(allowCreate: true)],
+    normalizationContext  : ['groups' => ['HiddenIdentifierDummy::out']],
+    denormalizationContext: ['groups' => ['HiddenIdentifierDummy::in']],
+    extraProperties       : [
+        'standard_put' => true,
+    ],
+)]
+#[Entity]
+class HiddenIdentifierDummy
+{
+    #[Id]
+    #[Column]
+    #[ApiProperty(identifier: false)]
+    public ?int $id = null;
+
+    #[Column]
+    #[ApiProperty(identifier: true)]
+    #[Groups(['HiddenIdentifierDummy::out'])]
+    public ?string $visibleId = null;
+
+    #[Column]
+    #[Groups(['HiddenIdentifierDummy::out', 'HiddenIdentifierDummy::in'])]
+    public string $foo = '';
+}
```

Comment:
can you explain your use case? I think that this should work already.

Reply:
I'll answer this first since it's faster  : i have a database with generated ids, and "business" ids as strings, which i want to use so API calls use the "business-recognisable" id. 
It lets users call "complex" IRIs (containing 2 or 3 such ids) without having to make multiple calls to get the right "internal" ids first.

And right now (last i tested at least) i can confirm it doesn't work : the orm id is not present in the uriVariables, and has no reason to be there, so this call to getReference() can't work : https://github.com/api-platform/core/blob/9b4b58ca0113a2645b58a116fe9e4bf200df8aa3/src/Doctrine/Common/State/PersistProcessor.php#L58

In my (only semi-confident) opinion uriVariables shouldn't be involved in this part of the framework at all, decoding the uri should be over and done by the time the workflow gets here.

Reply:
And its a good idea but this means that you have to decorate our processor to resolve the correct identifiers.

Let's take an example following your logic. Please note that I'm using a class that is not a Doctrine Entity as a resource.

```php
namespace App\ApiResource;
use ApiPlatform\Doctrine\Orm\State\Options;
use ApiPlatform\Metadata\ApiResource;

#[ApiResource(stateOptions: new Options(entityClass: SomeEntity::class))]
final class SomeApiResource {
  // name is "id" and gets automatically discovered by API Platform as an identifier, it is your business id
  public string $id; 
  // some API fields
  public string $name;
}
```

Now let's see that `SomeEntity`:

```php

namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity]
final class SomeEntity {
    #[ORM\Column(type: 'integer')]
    #[ORM\Id]
    #[ORM\GeneratedValue(strategy: 'AUTO')]
    public ?int $id = null;

    // This is our `id` from the SomeApiResource
    #[ORM\Column(type: 'text')]
    public string $myBusinessId;
}
```

We want users to do this kind of separation as it's cleaner. It's written in the [General Design Considerations](https://api-platform.com/docs/core/design/) and is considered best practice. 

Now you can see that we have an issue as API Platform will not be able to transform your `SomeApiResource::id` in the `SomeEntity::myBusinessId`. For that, we recommend to **decorate** our Processor/Providers and fix it yourself. This is exactly what we did by adding this `entityClass` option (see https://github.com/api-platform/core/blob/main/src/Doctrine/Orm/State/ItemProvider.php#L46). 

Maybe that we could add an `entityIdentifier` option that would use the `businessId` instead of the `id` to query stuff? 

I hope that this points you to a better direction then trying to make our processor work on complex resources that are also entities.